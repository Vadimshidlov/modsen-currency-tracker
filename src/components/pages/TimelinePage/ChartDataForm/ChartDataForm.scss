@use "@/constants/styles/fonts" as fonts;
@use "@/constants/styles/colors" as colors;

.chart-form__container {
  position: fixed;
  z-index: 2;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  background-color: rgba(0, 0, 0, 0.5);
  overflow: auto;
  transition: all 1s;

  .chart-form__content {
    padding-top: 35px;
    display: flex;
    flex-direction: column;
    align-items: center;
    position: relative;
    background-color: colors.$modal_background;
    margin: 10vh auto;
    max-width: 350px;
    border-radius: 10px;

    @media (max-width: 768px) {
      width: 270px;
    }

    @media (max-width: 420px) {
      width: 260px;
    }

    .chart-form__form {
      display: flex;
      flex-direction: column;
      align-items: center;
      row-gap: 10px;

      .chart-form__title {
        color: colors.$black_background;
        @include fonts.PoppinsLight;
        font-size: fonts.$font-size_25;
        margin-bottom: 10px;

        @media (max-width: 900px) {
          font-size: fonts.$font-size_20;
        }
      }

      .chart-form__field {
        .field_title {
          @include fonts.PoppinsLight;
          font-size: fonts.$font-size_20;

          @media (max-width: 900px) {
            font-size: fonts.$font-size_15;
          }
        }
      }


      .chart-form__submit {
        padding: 15px;
        background: transparent;
        color: colors.$color_white;
        @include fonts.PoppinsLight;
        font-size: fonts.$font-size_15;
        border: 2px solid colors.$color_grey__input-background;
        transition: all 1s;
        max-width: 200px;
        align-self: center;
        margin-top: 10px;
        margin-bottom: 10px;
        border-radius: 5px;

        &:hover {
          cursor: pointer;
          transform: scale(1.05);
        }

        @media (max-width: 768px) {
          padding: 7px;
          font-size: fonts.$font-size_15;
        }

        @media (max-width: 600px) {
          font-size: fonts.$font-size_12;
        }

        @media (max-width: 500px) {
          padding: 5px;
          font-size: fonts.$font-size_13;
        }
      }

      .chart-form__error {
        width: 60%;
        @include fonts.PoppinsLight;
        font-size: fonts.$font-size_15;
        text-align: center;
        color: colors.$chart-input__error-message;


        @media (max-width: 768px) {
          font-size: fonts.$font-size_13;
        }

        @media (max-width: 420px) {
          font-size: fonts.$font-size_12;
        }
      }
    }
  }

  .chart-form__close-button {
    background-color: colors.$modal_background;
    position: absolute;
    right: 10px;
    top: 10px;
    transition: all 1s;

    &:hover {
      cursor: pointer;
      transform: scale(1.3);
    }
  }
}




