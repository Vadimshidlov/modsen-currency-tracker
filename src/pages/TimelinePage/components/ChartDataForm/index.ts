export { default } from "@/pages/TimelinePage/components/ChartDataForm/ChartDataForm";
