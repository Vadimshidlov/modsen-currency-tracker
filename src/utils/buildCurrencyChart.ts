import Chart from "chart.js/auto";
import React from "react";
import { getRandomOhlcv } from "@/utils/getRandomOhlcv";

let chart: Chart;

export const buildCurrencyChart = (
    ref: React.RefObject<HTMLCanvasElement>,
    currencyValue: number,
    chartDate: string,
) => {
    // console.log(ref, "ref");
    console.log(currencyValue, chartDate, "buildCurrencyChart");

    return new Promise<void>((resolve, reject) => {
        try {
            const data = {
                datasets: [
                    {
                        // data: [...getRandomOhlcv(1000).datasets[0].data],
                        data: [
                            ...getRandomOhlcv(currencyValue, new Date(chartDate)).datasets[0].data,
                        ],
                        barPercentage: 0.5,
                        barThickness: 6,
                        maxBarThickness: 8,
                        minBarLength: 2,
                    },
                ],
            };

            const ctx = ref.current?.getContext("2d");
            // const ctx = ref.current?.getContext("2d");
            // ctx.clearRect(0, 0, ref.current?.width, ref.current?.height);

            if (ctx) {
                setTimeout(() => {
                    /* if (ref.current.v) {
                        ref.current.chart.destroy();
                    } */

                    // const chart = new Chart(ctx, {
                    if (chart) {
                        // chart.destroy();
                        chart.update();
                    }

                    chart = new Chart(ctx, {
                        type: "candlestick",
                        data,
                        options: {
                            animations: {
                                tension: {
                                    duration: 1000,
                                    easing: "linear",
                                    from: 1,
                                    to: 0,
                                    loop: false,
                                },
                            },
                            responsive: true,
                            tooltips: {
                                mode: "index",
                            },
                            interaction: {
                                intersect: false,
                            },
                            plugins: {
                                title: {
                                    display: true,
                                    text: "Modsen Currency Tracker",
                                },
                                legend: {
                                    display: false,
                                },
                                tooltip: {
                                    intersect: false,
                                    position: "nearest",
                                },
                            },
                            scales: {
                                y: {
                                    display: true,
                                    title: {
                                        display: true,
                                        text: "Value",
                                        position: "bottom",
                                    },
                                    tooltip: {},
                                    position: "right",
                                },
                                xAxis: {
                                    type: "time",
                                    time: {
                                        unit: "day",
                                        tooltipFormat: "YYYY-MM-DD",
                                        displayFormats: {
                                            day: "DD",
                                        },
                                    },
                                    title: {
                                        display: true,
                                        text: "Day",
                                        position: "bottom",
                                    },
                                    barPercentage: 0.5,
                                    barThickness: 1,
                                    grid: {
                                        display: true,
                                        drawOnChartArea: true,
                                        drawTicks: true,
                                    },
                                },
                            },
                        },
                    });
                    // chart.render();
                    // resolve(chart.update());
                    // resolve(chart.render());

                    // chart.render();
                    // chart.render();
                    console.log(chart);

                    // chart.destroy();
                    chart.render();
                    // successChartBuildMessage();
                    resolve();
                }, 1000);
            } else {
                reject(new Error("Could not get canvas context"));
            }
        } catch (error) {
            reject(error);
        }
    });
};
