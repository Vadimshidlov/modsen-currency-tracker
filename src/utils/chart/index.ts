export { buildCurrencyChart } from "@/utils/chart/buildCurrencyChart";
export { getRandomOhlcv } from "@/utils/chart/getRandomOhlcv";
export { getDataForChart } from "@/utils/chart/getDataForChart";
export { validateChartFormSubmit } from "@/utils/chart/validateChartFormSubmit";
