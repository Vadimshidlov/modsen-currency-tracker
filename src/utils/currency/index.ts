export { filterCurrency } from "@/utils/currency/filterCurrency";
export { getCurrencyIcon } from "@/utils/currency/getCurrencyIcon";
export { getSecondCurrency } from "@/utils/currency/getSecondCurrency";
export { getConvertResult } from "@/utils/currency/getConvertResult";
export { getCurrencyTitleWithCode } from "@/utils/currency/getCurrencyTitleWithCode";
export { getCurrencyValueWithCode } from "@/utils/currency/getCurrencyValueWithCode";
