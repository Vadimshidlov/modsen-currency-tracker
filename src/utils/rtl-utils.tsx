import React, { PropsWithChildren } from "react";
import { Store, applyMiddleware, createStore } from "redux";
import { thunk } from "redux-thunk";
import { render } from "@testing-library/react";
import type { RenderOptions } from "@testing-library/react";
import { Provider } from "react-redux";
import { RootStateType, rootReducer } from "@/store/reducers";

interface ExtendedRenderOptions extends Omit<RenderOptions, "queries"> {
    preloadedState?: Partial<RootStateType>;
    store?: Store;
}

export function renderWithProviders(
    ui: React.ReactElement,
    {
        preloadedState = {},
        store = createStore(rootReducer, preloadedState, applyMiddleware(thunk)),
        ...renderOptions
    }: ExtendedRenderOptions = {},
) {
    function Wrapper({ children }: PropsWithChildren<object>): JSX.Element {
        return <Provider store={store}>{children}</Provider>;
    }
    return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

export * from "@testing-library/react";

export { renderWithProviders as render };
