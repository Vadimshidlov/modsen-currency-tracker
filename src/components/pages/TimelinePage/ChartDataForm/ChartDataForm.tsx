import "@/components/pages/HomePage/ModalWindow/ModalWindow.scss";
import { createPortal } from "react-dom";
import React, { ChangeEvent, Component, FormEvent } from "react";
import CloseModalIcon from "@/assets/svg/modalWindow/icons8-close.svg";
import Text from "@/components/pages/TimelinePage/Text/Text";
import "@/components/pages/TimelinePage/ChartDataForm/ChartDataForm.scss";
import TextInput from "@/components/pages/BankCard/TextInput/TextInput";
import { validateChartFormSubmit } from "@/utils/chart/validateChartFormSubmit";
import { ChartDataFormPropsType, ChartDataFormStateType } from "@/types/TimeLinePageTypes/types";

const VALIDATE_ERROR_MESSAGE =
    "High must be higher than low value, close and open must be between high and low values";

class ChartDataForm extends Component<ChartDataFormPropsType, ChartDataFormStateType> {
    constructor(props: ChartDataFormPropsType) {
        super(props);

        this.state = {
            highInputValue: "",
            closeInputValue: "",
            openInputValue: "",
            lowInputValue: "",
            submitError: "",
        };

        this.onSubmitHandler = this.onSubmitHandler.bind(this);
        this.onFocusHandler = this.onFocusHandler.bind(this);
    }

    onSubmitHandler(e: FormEvent) {
        e.preventDefault();
        const { onFormSubmit, onClose } = this.props;

        const { lowInputValue, openInputValue, highInputValue, closeInputValue } = this.state;

        if (
            validateChartFormSubmit(lowInputValue, openInputValue, highInputValue, closeInputValue)
        ) {
            onFormSubmit(+highInputValue, +closeInputValue, +openInputValue, +lowInputValue);
            onClose();
        } else {
            this.setState({
                submitError: VALIDATE_ERROR_MESSAGE,
            });
        }
    }

    onFocusHandler() {
        this.setState({
            submitError: "",
        });
    }

    render() {
        const { closeInputValue, lowInputValue, openInputValue, highInputValue, submitError } =
            this.state;
        const { onClose } = this.props;

        return createPortal(
            <div className="chart-form__container">
                <div className="chart-form__content">
                    <form className="chart-form__form" onSubmit={this.onSubmitHandler}>
                        <Text className="chart-form__title">Change data for chart</Text>

                        <div className="chart-form__field field">
                            <Text className="field_title">High:</Text>
                            <TextInput
                                className="currency__input"
                                value={highInputValue}
                                onChange={(e: ChangeEvent<HTMLInputElement>) => {
                                    const { value } = e.target;

                                    this.setState({
                                        highInputValue: value,
                                    });
                                }}
                                onFocus={this.onFocusHandler}
                            />
                        </div>
                        <div className="chart-form__field">
                            <Text className="field_title">Close:</Text>
                            <TextInput
                                className="currency__input"
                                value={closeInputValue}
                                onChange={(e: ChangeEvent<HTMLInputElement>) => {
                                    const { value } = e.target;

                                    this.setState({
                                        closeInputValue: value,
                                    });
                                }}
                                onFocus={this.onFocusHandler}
                            />
                        </div>
                        <div className="chart-form__field">
                            <Text className="field_title">Open:</Text>
                            <TextInput
                                className="currency__input"
                                value={openInputValue}
                                onChange={(e: ChangeEvent<HTMLInputElement>) => {
                                    const { value } = e.target;

                                    this.setState({
                                        openInputValue: value,
                                    });
                                }}
                                onFocus={this.onFocusHandler}
                            />
                        </div>
                        <div className="chart-form__field">
                            <Text className="field_title">Low:</Text>
                            <TextInput
                                className="currency__input"
                                value={lowInputValue}
                                onChange={(e: ChangeEvent<HTMLInputElement>) => {
                                    const { value } = e.target;

                                    this.setState({
                                        lowInputValue: value,
                                    });
                                }}
                                onFocus={this.onFocusHandler}
                            />
                        </div>
                        {submitError && <Text className="chart-form__error">{submitError}</Text>}
                        <button className="chart-form__submit">Update</button>
                    </form>

                    <CloseModalIcon
                        className="chart-form__close-button"
                        width={25}
                        height={25}
                        onClick={onClose}
                    />
                </div>
            </div>,
            document.getElementById("portal"),
        );
    }
}

export default ChartDataForm;
