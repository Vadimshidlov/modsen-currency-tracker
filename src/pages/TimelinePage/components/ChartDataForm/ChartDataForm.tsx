import "@/pages/HomePage/components/ModalWindow/ModalWindow.scss";
import React, { ChangeEvent, Component, FormEvent } from "react";
import CloseModalIcon from "@/assets/svg/modalWindow/icons8-close.svg";
import Text from "@/components/Text/index";
import "@/pages/TimelinePage/components/ChartDataForm/ChartDataForm.scss";
import TextInput from "@/pages/BankCard/components/TextInput/index";
import { validateChartFormSubmit } from "@/utils/chart/index";
import { ChartDataFormPropsType, ChartDataFormStateType } from "@/types/TimeLinePageTypes";
import { Portal } from "@/components/Portal/index";
import BlurBackground from "@/components/BlurBackground/index";

const VALIDATE_ERROR_MESSAGE =
    "High must be higher than low value, close and open must be between high and low values";

class ChartDataForm extends Component<ChartDataFormPropsType, ChartDataFormStateType> {
    constructor(props: ChartDataFormPropsType) {
        super(props);

        this.state = {
            highInputValue: "",
            closeInputValue: "",
            openInputValue: "",
            lowInputValue: "",
            submitError: "",
        };

        this.handleFormSubmit = this.handleFormSubmit.bind(this);
        this.handleInputFocus = this.handleInputFocus.bind(this);
    }

    componentDidMount() {
        document.body.classList.add("modal-open");
    }

    componentWillUnmount() {
        document.body.classList.remove("modal-open");
    }

    handleFormSubmit(e: FormEvent) {
        e.preventDefault();
        const { onFormSubmit, onClose } = this.props;

        const { lowInputValue, openInputValue, highInputValue, closeInputValue } = this.state;

        if (
            validateChartFormSubmit(lowInputValue, openInputValue, highInputValue, closeInputValue)
        ) {
            onFormSubmit(+highInputValue, +closeInputValue, +openInputValue, +lowInputValue);
            onClose();

            return;
        }

        this.setState({
            submitError: VALIDATE_ERROR_MESSAGE,
        });
    }

    handleInputFocus() {
        this.setState({
            submitError: "",
        });
    }

    render() {
        const { closeInputValue, lowInputValue, openInputValue, highInputValue, submitError } =
            this.state;
        const { onClose } = this.props;

        return (
            <Portal>
                <div className="chart-form__container">
                    <div className="chart-form__content">
                        <form className="chart-form__form" onSubmit={this.handleFormSubmit}>
                            <Text className="chart-form__title">Change data for chart</Text>

                            <div className="chart-form__field field">
                                <Text className="field_title">High:</Text>
                                <TextInput
                                    className="currency__input"
                                    data-testid="chart-input-high"
                                    value={highInputValue}
                                    onChange={(e: ChangeEvent<HTMLInputElement>) => {
                                        const { value } = e.target;

                                        this.setState({
                                            highInputValue: value,
                                        });
                                    }}
                                    onFocus={this.handleInputFocus}
                                />
                            </div>
                            <div className="chart-form__field">
                                <Text className="field_title">Close:</Text>
                                <TextInput
                                    className="currency__input"
                                    data-testid="chart-input-close"
                                    value={closeInputValue}
                                    onChange={(e: ChangeEvent<HTMLInputElement>) => {
                                        const { value } = e.target;

                                        this.setState({
                                            closeInputValue: value,
                                        });
                                    }}
                                    onFocus={this.handleInputFocus}
                                />
                            </div>
                            <div className="chart-form__field">
                                <Text className="field_title">Open:</Text>
                                <TextInput
                                    className="currency__input"
                                    data-testid="chart-input-open"
                                    value={openInputValue}
                                    onChange={(e: ChangeEvent<HTMLInputElement>) => {
                                        const { value } = e.target;

                                        this.setState({
                                            openInputValue: value,
                                        });
                                    }}
                                    onFocus={this.handleInputFocus}
                                />
                            </div>
                            <div className="chart-form__field">
                                <Text className="field_title">Low:</Text>
                                <TextInput
                                    className="currency__input"
                                    data-testid="chart-input-low"
                                    value={lowInputValue}
                                    onChange={(e: ChangeEvent<HTMLInputElement>) => {
                                        const { value } = e.target;

                                        this.setState({
                                            lowInputValue: value,
                                        });
                                    }}
                                    onFocus={this.handleInputFocus}
                                />
                            </div>
                            {submitError && (
                                <Text className="chart-form__error">{submitError}</Text>
                            )}
                            <button className="chart-form__submit" data-testid="chart-submit">
                                Update
                            </button>
                        </form>

                        <CloseModalIcon
                            className="chart-form__close-button"
                            width={25}
                            height={25}
                            onClick={onClose}
                        />
                    </div>
                    <BlurBackground onClick={onClose} />
                </div>
            </Portal>
        );
    }
}

export default ChartDataForm;
