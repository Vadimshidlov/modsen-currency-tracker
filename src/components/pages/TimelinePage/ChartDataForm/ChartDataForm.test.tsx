import { act, fireEvent, waitFor } from "@testing-library/react";
import ChartDataForm from "@/components/pages/TimelinePage/ChartDataForm/ChartDataForm";
import { renderWithProviders, screen } from "@/utils/rtl-utils";

describe("ChartDataForm component tests", () => {
    let container: HTMLDivElement;

    beforeEach(() => {
        container = document.createElement("div");
        container.id = "portal";
        document.body.appendChild(container);
    });

    afterEach(() => {
        document.body.removeChild(container);
        container = null;
    });

    it("renders form with correct fields", async () => {
        await act(async () => {
            await waitFor(() => {
                renderWithProviders(<ChartDataForm onFormSubmit={() => {}} onClose={() => {}} />, {
                    container,
                });
            });
        });

        expect(screen.getByText("Change data for chart")).toBeInTheDocument();
        expect(screen.getByTestId("chart-input-high")).toBeInTheDocument();
        expect(screen.getByTestId("chart-input-close")).toBeInTheDocument();
        expect(screen.getByTestId("chart-input-open")).toBeInTheDocument();
        expect(screen.getByTestId("chart-input-low")).toBeInTheDocument();
        expect(screen.getByTestId("chart-submit")).toBeInTheDocument();
    });

    it("calls onFormSubmit when form is submitted with valid data", () => {
        const onFormSubmit = jest.fn();
        const onClose = jest.fn();

        renderWithProviders(<ChartDataForm onFormSubmit={onFormSubmit} onClose={onClose} />);

        const highInput = screen.getByTestId("chart-input-high");
        const closeInput = screen.getByTestId("chart-input-close");
        const openInput = screen.getByTestId("chart-input-open");
        const lowInput = screen.getByTestId("chart-input-low");
        const submitButton = screen.getByTestId("chart-submit");

        fireEvent.change(highInput, { target: { value: "100" } });
        fireEvent.change(closeInput, { target: { value: "90" } });
        fireEvent.change(openInput, { target: { value: "95" } });
        fireEvent.change(lowInput, { target: { value: "85" } });

        fireEvent.click(submitButton);

        expect(onFormSubmit).toHaveBeenCalled();
        expect(onFormSubmit).toHaveBeenCalledWith(100, 90, 95, 85);
        expect(onClose).toHaveBeenCalled();
    });

    it("displays error message when form is submitted with invalid data", () => {
        const onFormSubmit = jest.fn();
        const onClose = jest.fn();

        renderWithProviders(<ChartDataForm onFormSubmit={onFormSubmit} onClose={onClose} />);

        const submitButton = screen.getByTestId("chart-submit");

        fireEvent.click(submitButton);
        expect(
            screen.getByText(
                "High must be higher than low value, close and open must be between high and low values",
            ),
        ).toBeInTheDocument();
    });

    it("clears error message onFocus", () => {
        const onFormSubmit = jest.fn();
        const onClose = jest.fn();

        renderWithProviders(<ChartDataForm onFormSubmit={onFormSubmit} onClose={onClose} />);

        const highInput = screen.getByTestId("chart-input-high");

        fireEvent.change(highInput, { target: { value: "100" } });

        expect(
            screen.queryByText(
                "High must be higher than low value, close and open must be between high and low values",
            ),
        ).not.toBeInTheDocument();
    });

    it("snapshot matches", async () => {
        const onFormSubmit = jest.fn();
        const onClose = jest.fn();

        const { asFragment } = renderWithProviders(
            <ChartDataForm onFormSubmit={onFormSubmit} onClose={onClose} />,
        );
        expect(asFragment()).toMatchSnapshot();
    });
});
