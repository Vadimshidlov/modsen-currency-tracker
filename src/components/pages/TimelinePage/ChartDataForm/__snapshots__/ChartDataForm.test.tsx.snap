// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChartDataForm component tests snapshot matches 1`] = `<DocumentFragment />`;
